File: core/rules/tw_nhi.py
"""健保卡號（TW_NHI）格式驗證與偵測（無 checksum，僅驗證格式）。"""

import re

# 候選字串：連續 12 碼數字，前後不可緊接英數字，避免截斷更長的字串
_TW_NHI_PATTERN = re.compile(r"(?<![A-Za-z0-9])\d{12}(?![A-Za-z0-9])")


def is_valid_tw_nhi(nhi_str: str) -> bool:
    """驗證字串是否為合法健保卡號格式（純 12 碼數字，不驗 checksum）。"""
    if not isinstance(nhi_str, str):
        return False

    candidate = nhi_str.strip()
    return len(candidate) == 12 and candidate.isdigit()


def detect_tw_nhi(text: str) -> dict:
    """在文字中找出所有健保卡號，回傳符合 docs/interface.md 格式的偵測結果。"""
    spans = []
    count = 0
    for match in _TW_NHI_PATTERN.finditer(text):
        candidate = match.group()
        if not is_valid_tw_nhi(candidate):
            continue
        count += 1
        spans.append({
            "start": match.start(),
            "end": match.end(),
            "type": "TW_NHI",
            "text": candidate,
            "confidence": 0.6,
            "source": "rule",
            "replacement": f"[TW_NHI_{count}]",
        })
    return {"text": text, "spans": spans}

File: extension/src/core/tw_nhi.ts
/**
 * 健保卡號（TW_NHI）格式驗證與偵測（無 checksum，僅驗證格式）。
 * 移植自 Python 版 `core/rules/tw_nhi.py`，行為必須完全一致。
 */

import type { DetectionResult, Span } from './types';
import { findIter, isAsciiDigits, makeRuleSpan } from './util';

/** 候選字串：連續 12 碼數字，前後不可緊接英數字，避免截斷更長的字串。 */
const TW_NHI_PATTERN = /(?<![A-Za-z0-9])[0-9]{12}(?![A-Za-z0-9])/g;

/** 驗證字串是否為合法健保卡號格式（純 12 碼數字，不驗 checksum）。 */
export function isValidTwNhi(nhiStr: unknown): boolean {
  if (typeof nhiStr !== 'string') return false;
  const candidate = nhiStr.trim();
  return candidate.length === 12 && isAsciiDigits(candidate);
}

/** 在文字中找出所有健保卡號。 */
export function detectTwNhi(text: string): DetectionResult {
  const spans: Span[] = [];
  let count = 0;
  for (const match of findIter(TW_NHI_PATTERN, text)) {
    const candidate = match[0];
    if (!isValidTwNhi(candidate)) continue;
    count += 1;
    spans.push(
      makeRuleSpan(match.index, match.index + candidate.length, 'TW_NHI', candidate, 0.6, count),
    );
  }
  return { text, spans };
}
